packages/upload/src/vaadin-upload.js

```diff
@@ -795,16 +816,17 @@ class Upload extends ElementMixin(ThemableMixin(ControllerMixin(PolymerElement))
       if (file.xhr) {
         file.xhr.abort();
       }
-      this._notifyFileChanges(file);
+
+      this._updateFileList();
     }
   }
 
   /** @private */
-  _notifyFileChanges(file) {
-    const p = `files.${this.files.indexOf(file)}.`;
-    Object.keys(file).forEach((i) => {
-      this.notifyPath(p + i, file[i]);
-    });
+  _updateFileList() {
+    const files = [...this.files];
+    // Re-render file list DOM without re-assigning `files`
+    // to avoid dispatching `files-changed` notify event.
+    this._files = files;
```

Comment:
Naming is getting pretty confusing with:
- `files`
- `_files`
- `_filesChanged` (which one?)
- `_updateFileList`
- `__updateFileList`

How about just calling `requestContentUpdate` on the file list when a property of a file changes? That would be more straight-forward than changing the items reference just to trigger the observer.

Alternatively we could simplify the observer setup like this: https://github.com/vaadin/web-components/commit/265cc28df358c01df783641a90cf9b08080746aa
This changes `__updateFileList` to always pass a new items reference to the file list, which should be fine as by that point we already know that something has definitely changed and the file list needs to be re-rendered.

I think I would prefer the first option. It would circumvent the debouncer in the file list, though I'm not sure if it's actually needed here.

Reply:
I agree, this looks quite confusing. Calling `requestContentUpdate()` is indeed a better idea, will do that.
Having a single observer method but also calling it manually is something I'd like to avoid.

Reply:
Done.